fix(zone.js): should allow add passive/non-passive listeners together

JiaLiPassion opened the pull request:
Close #45020
In the current version, if we add both `passive` and `not passive` listeners for the same eventName together, they will be registered with all passive or all non passive listeners depends on the order.

```
import 'zone.js';
div1.addEventListener('mousemove', (ev) => {}, { passive: true });
div1.addEventListener('mousemove', (ev) => {
ev.preventDefault(); // throws error since this one is also be registered as a passive event handler
});

div2.addEventListener('mousemove', (ev) => {
});
div2.addEventListener('mousemove', (ev) => {
ev.preventDefault(); // will not throw error since this one is also be registered as non passive event handler
}, { passive: true });

```
So this PR fix this issue and allow both passive and non-passive listeners registeration together whatever the order.

JeanMeche commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -434,6 +446,21 @@ export function patchEventTarget(
+        if (passiveSupported && existingTasks.length) {
+          // check whether all tasks are passive.
+          let foundNonPassive = false;
+          for (let i = 0; i < existingTasks.length; i++) {
```
nit: could be a `existingTasks.find() !== undefined` ...

JiaLiPassion replied:
Thank you, I have a wrong memory about performance comparsion, `Array.prototype.find()` is faster, I will update this part.

AndrewKushnir commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -110,6 +110,10 @@ export function patchEventTarget(
       return;
     }
+    const options = task.options;
+    if (typeof options === 'object' && options?.passive === true) {
```
This check is repeated multiple times in this file, I think it'd be great to extract it to a separate function (which would also act as a self-documenting code), for example:

```
function isPassiveEventListener(options: {}): boolean {
  return typeof options === 'object' && options?.passive === true;
}
```
after that the code may look like:
```suggestion
    if (isPassiveEventListener(options)) {
```

JiaLiPassion replied:
Thanks @AndrewKushnir , sure, updated.

AndrewKushnir commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -110,6 +110,10 @@ export function patchEventTarget(
     }
+    const options = task.options;
+    if (typeof options === 'object' && options?.passive === true) {
+      return;
```
Could you please add a quick comment into the code on why we exit early?

JiaLiPassion replied:
Sure. added. Thanks.

JiaLiPassion commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -434,6 +446,12 @@ export function patchEventTarget(
         }
+        // check whether all tasks are passive.
+        if (passiveSupported && existingTasks.length &&
+            !existingTasks.find(
```
@alfaproject , yeah, just updated, thanks.

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -104,12 +104,25 @@ export function patchEventTarget(
   const PREPEND_EVENT_LISTENER = 'prependListener';
   const PREPEND_EVENT_LISTENER_SOURCE = '.' + PREPEND_EVENT_LISTENER + ':';
 
+  function isPassiveEventListener(options: any): boolean {
```
```suggestion
  function isPassiveEventListener(options: unknown): boolean {
```

alan-agius4 reviewed (approved):
One NIT

alan-agius4 commented:
[TGP](https://test.corp.google.com/OCL:590919640:BASE:590919576:1702564213551:51fa8948)

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -104,12 +104,25 @@ export function patchEventTarget(
   const PREPEND_EVENT_LISTENER = 'prependListener';
```
This change breaks a large number of apps in Google. 

Error: 
```
TypeError: task.callback.apply is not a function
    at _ZoneDelegate.invokeTask (third_party/javascript/angular2/rc/packages/zone.js/lib/zone.ts?l=1187
```

JiaLiPassion committed:
fix(zone.js): should allow add passive/non-passive listeners together

In the current version, if we add both `passive` and `not passive` listeners for the
same eventName together, they will be registered with all passive or all non passive
listeners depends on the order.

```
import 'zone.js';
div1.addEventListener('mousemove', (ev) => {}, { passive: true });
div1.addEventListener('mousemove', (ev) => {
ev.preventDefault(); // throws error since this one is also be registered as a passive event handler
});

div2.addEventListener('mousemove', (ev) => {
});
div2.addEventListener('mousemove', (ev) => {
ev.preventDefault(); // will not throw error since this one is also be registered as non passive event handler
}, { passive: true });

```
So this PR fix this issue and allow both passive and non-passive listeners registeration
together whatever the order.

PR closes #45020

alan-agius4 commented:
[TGP](https://test.corp.google.com/OCL:591655276:BASE:591655206:1702812413340:4a6322b4)

alan-agius4 commented on packages/zone.js/lib/common/events.ts:
```diff
@@ -391,9 +391,33 @@ export function patchEventTarget(
-            passiveSupported && !!passiveEvents && passiveEvents.indexOf(eventName) !== -1;
-        const options = buildEventListenerOptions(arguments[2], passive);
+        const isPassiveEvent =
+            passiveSupported && passiveEvents && passiveEvents.indexOf(eventName) !== -1;
```
NIT

```suggestion
            passiveSupported && passiveEvents && passiveEvents.includes(eventName);
```

JiaLiPassion replied:
thanks, updated.

alan-agius4 commented:
[TGP](https://test.corp.google.com/OCL:591655276:BASE:592132091:1702973244330:2f0a611d)
